add CC100

erip opened the pull request:
Closes #1494 

cc @parmeet I've had some success with this locally...

erip commented on torchtext/datasets/cc100.py:
```diff
@@ -0,0 +1,57 @@
+
+DATASET_NAME = "CC100"
+
+
```
This may be a bit of a pain, but should we audit the num lines per language or does it make better sense to have "non-standard" documentation for this "non-standard" dataset? 😄

erip commented:
Calling this a WIP is maybe a bit conservative. It works locally, but it seems like testing it requires some additional thought because this isn't a traditional supervised dataset (so it doesn't really have a concept of splits). I welcome discussion here and will leave the WIP until we've settled on something.

abhinavarora commented:
Looks like the server hosting the dataset is back up :-)

erip commented:
>  it doesn't really have a concept of splits

Neither does EnWiki9, so I followed that approach. 😄

erip commented:
It seems like `*args` don't get propagated appropriately in the split-wrapper decorator. I can give the `language_code` a default value (or make it a `Tuple[str]` to handle multiple values).

parmeet commented:
Thanks @erip for taking up this one. I also have a local commit and was actually waiting for resolving the server issue first before check-in. Let's work through your PR :)

erip commented:
I have incorporated the feedback from your review, @parmeet, but I'm observing some weird behavior... For whatever reason the `map` call after readlines isn't working as expected... 

```python
>>> from torchtext.datasets import CC100
>>> dp = CC100(language_code="zu")
/Users/erip/opt/miniconda3/envs/torchtext-dev/lib/python3.8/site-packages/torch/utils/data/datapipes/utils/common.py:24: UserWarning: Lambda function is not supported for pickle, please use regular python function or functools.partial instead.
  warnings.warn(
>>> next(iter(dp))
<torch.utils.data.datapipes.iter.callable.MapperIterDataPipe object at 0x7f96d1c305e0>
>>> next(iter(next(iter(dp))))
('zu', 'Ngiyabonga wathola: 6')
```

erip commented on torchtext/datasets/cc100.py:
```diff
@@ -0,0 +1,61 @@
+        filepath_fn=lambda x: os.path.join(root, os.path.basename(x).rstrip(".xz"))
+    )
+    cache_decompressed_dp = FileOpener(cache_decompressed_dp, mode="b").read_from_xz()
+    cache_decompressed_dp = cache_decompressed_dp.filter(lambda x: os.path.basename(x).rstrip(".xz") in x[0])
```
Good point, done!

erip commented:
I've incorporated your feedback, @parmeet. It seems like there's some interference with `_wrap_split_argument`. I'm trying to mimic the behavior in `EnWiki9` but when I give `CC100` no default args, I encounter this error:

```python
>>> from torchtext.datasets import CC100
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/erip/Code/text/torchtext/__init__.py", line 7, in <module>
    from . import datasets
  File "/Users/erip/Code/text/torchtext/datasets/__init__.py", line 6, in <module>
    from .cc100 import CC100
  File "/Users/erip/Code/text/torchtext/datasets/cc100.py", line 36, in <module>
    def CC100(root: str, split: Union[Tuple[str], str], language_code: str):
  File "/Users/erip/Code/text/torchtext/data/datasets_utils.py", line 309, in new_fn
    return _wrap_split_argument_with_fn(fn, splits)
  File "/Users/erip/Code/text/torchtext/data/datasets_utils.py", line 301, in _wrap_split_argument_with_fn
    new_sig = new_sig.replace(parameters=tuple(new_params))
  File "/Users/erip/opt/miniconda3/envs/torchtext-dev/lib/python3.8/inspect.py", line 2877, in replace
    return type(self)(parameters,
  File "/Users/erip/opt/miniconda3/envs/torchtext-dev/lib/python3.8/inspect.py", line 2810, in __init__
    raise ValueError(msg)
ValueError: non-default argument follows default argument
```

and when I give default values, the returned value is _always_ a tuple (even though there's one split). Does this seem familiar?

erip commented:
Ah, I figured it out... Either everything needs a default value or only arguments after `split` can have a default value. This seems good now and should be consistent with other datasets. 👍

parmeet reviewed (approved):
Thanks @erip for contributing CC-100, this is the first large-scale dataset added to the repo :).  Overall LGTM! As a follow-up we potentially want to figure out what to do with MD5 checks. I guess, It is not a trivial exercise to calculate MD5 for all 100+ huge files :(
